core/modules/replica/SqlRequest.h

```diff
@@ -46,57 +48,173 @@ namespace qserv {
+     *   Laving the default value of the parameter to 0 will result in not imposing any
+     *   explicit restrictions on a size of the result set. NOte that other, resource-defined
+     *   restrictions will still apply. The later includes the maximum size of the Google Protobuf
+     *   objects, the amount of available memory, etc.
+
+     * @param priority
+     *   priority level of the request
+     *
+     * @param keepTracking
+     *   keep tracking the request before it finishes or fails
+     *
+     * @param messenger
+     *   interface for communicating with workers
      */
-    SqlResultSet const& responseData() const;
+    SqlBaseRequest(ServiceProvider::Ptr const& serviceProvider,
+                   boost::asio::io_service& io_service,
+                   std::string const& worker,
+                   uint64_t maxRows,
+                   int priority,
+                   bool keepTracking,
+                   std::shared_ptr<Messenger> const& messenger);
+
+    /// @see Request::startImpl()
+    void startImpl(util::Lock const& lock) final;
+
+    /// @see Request::savePersistentState()
+    void savePersistentState(util::Lock const& lock) final;
+
+    /**
+     * Request body to be sent to the worker. The content of the request is partially
+     * set by this class's constructor, and it's fully initialized by the constructor
+     * of a subclass depending on a type of request.
+     */
+    ProtocolRequestSql requestBody;
+
+private:
+
+    /**
+     * Start the timer before attempting the previously failed
+     * or successful (if a status check is needed) step.
+     *
+     * @param lock
+     *   a lock on Request::_mtx must be acquired before calling this method
+     */
+    void _wait(util::Lock const& lock);
+
+    /**
+     * Callback handler for the asynchronous operation
+     *
+     * @param ec
+     *   error code to be checked
+     */
+    void _awaken(boost::system::error_code const& ec);
+
+    /**
+     * Send the serialized content of the buffer to a worker
+     *
+     * @param lock
+     *   a lock on Request::_mtx must be acquired before calling this method
+     */
+    void _send(util::Lock const& lock);
+
+    /**
+     * Process the completion of the requested operation
+     *
+     * @param success
+     *   'true' indicates a successful response from a worker
+     *
+     * @param response
+     *   response from a worker (if success)
+     */
+    void _analyze(bool success,
+                  ProtocolResponseSql const& response);
+
+    /// Request-specific parameters of the target request
+    SqlRequestParams _targetRequestParams;
+
+    /// The results reported by a worker service
+    SqlResultSet _responseData;
+};
+
+
+/**
+ * Class SqlQueryRequest represents Controller-side requests for initiating
+ * arbitrary database queries at a remote worker nodes.
+ */
+class SqlQueryRequest : public SqlBaseRequest  {
```

Comment:
when should this class get its own file?

Reply:
(and so on with the other classes in this file...)

Reply:
Agreed. I will refactor this huge file into many (one file per class).